receiver/receiverhelper/receiver.go

```diff
@@ -0,0 +1,265 @@
+		if start != nil {
+			if err := start(ctx, host); err != nil {
+				return err
+			}
+		}
+
+		if err := mr.initializeScrapers(ctx); err != nil {
+			return err
+		}
+
+		mr.startScraping()
+		return nil
+	}
+
+	// wrap the shutdown function with a call to close scrapers
+	// and stop scraping
+	shutdown := mr.shutdown
+	mr.shutdown = func(ctx context.Context) error {
+		mr.stopScraping()
+
+		var errors []error
+
+		if err := mr.closeScrapers(ctx); err != nil {
+			errors = append(errors, err)
+		}
+
+		if shutdown != nil {
+			if err := shutdown(ctx); err != nil {
+				errors = append(errors, err)
+			}
+		}
+
+		return componenterror.CombineErrors(errors)
+	}
+
+	return mr, nil
+}
+
+// initializeScrapers initializes all the scrapers
+func (mr *metricsReceiver) initializeScrapers(ctx context.Context) error {
+	for _, scraper := range mr.scrapers {
+		if scraper.initialize == nil {
+			continue
+		}
+
+		if err := scraper.initialize(ctx); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// startScraping initiates a ticker that calls Scrape based on the configured
+// collection interval.
+func (mr *metricsReceiver) startScraping() {
+	// TODO1: use one ticker for each set of scrapers that have the same collection interval.
+	// TODO2: consider allowing different "Scrape" functions to be configured, i.e. functions
+	//        that return MetricsSlice or ResourceMetricsSlice (similar to the existing Scraper
+	//        & ResourceScraper interfaces in the host metrics receiver). That will allow data
+	//        from multiple scrapers (that have the same collection interval) to be batched.
+
+	for i := 0; i < len(mr.scrapers); i++ {
+		scraper := mr.scrapers[i]
+		go func() {
+			collectionInterval := mr.defaultCollectionInterval
+			if scraper.collectionInterval != 0 {
+				collectionInterval = scraper.collectionInterval
+			}
+
+			ticker := time.NewTicker(collectionInterval)
+			defer ticker.Stop()
+
+			for {
+				select {
+				case <-ticker.C:
+					mr.scrapeAndReport(context.Background(), scraper)
+				case <-mr.done:
+					return
+				}
+			}
+		}()
+	}
+}
+
+// scrapeAndReport calls the Scrape function of the provided Scraper, records
+// observability information, and passes the scraped metrics to the next component.
+func (mr *metricsReceiver) scrapeAndReport(ctx context.Context, scraper *scraper) {
+	// TODO: Add observability metrics support
+	metrics, err := scraper.scrape(ctx)
+	if err != nil {
+		return
+	}
+
+	mr.nextConsumer.ConsumeMetrics(ctx, metrics)
+}
+
+// stopScraping stops the ticker
+func (mr *metricsReceiver) stopScraping() {
+	close(mr.done)
+}
+
+// closeScrapers closes all the scrapers
+func (mr *metricsReceiver) closeScrapers(ctx context.Context) error {
+	var errors []error
+
+	for _, scraper := range mr.scrapers {
+		if scraper.close == nil {
+			continue
+		}
+
+		if err := scraper.close(ctx); err != nil {
+			errors = append(errors, err)
+		}
+	}
+
+	return componenterror.CombineErrors(errors)
+}
```

Comment:
Maybe we would want to add / extend `NewTraceReceiver`, `NewMetricReceiver`, & `NewLogsReceiver` functions here at some point that implement standard HTTP and/or gRPC servers for "push" type receivers.

I haven't implemented any of those types of receivers so I'm not sure if it would be straightforward to factor out that code, or worth the effort. For now, these helper functions are only intended to support the `NewMetricReceiver` function where you are configuring at least one Scraper, i.e. for "pull" type receivers.